### PR TITLE
adding unittests

### PR DESCRIPTION
**Pull request by alextsaihi**
Emre seems to be adding a larger number of changes so he thought it would be best to divide his work and mine into two PRs

- unittests for validate_response.py and report_runner.py

### CODE REVIEW

**Comment by emre-f** [comment]
@alextsaihi  is it expected to fail the unit tests in the GitHub Actions portion?

File: unittests/resources/bad_service_info_2.json
Comment (emre-f) [inline unittests/resources/bad_service_info_2.json:1]: @alextsaihi what is the purpose of this second `bad_service_info` JSON?

File: unittests/v1.2.0/test_drs_objects.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+# CONSTANTS
+schema_name = "v1.2.0/drs_object.json"
+test_description = "testing drs-object"
+
+# spec has required fields
+# unittest tests if cs is testing correctly
+# check validate schema code runs bad drs object, expect a fail or error from cs
+
+# import drs_object.json
```
Comment (emre-f) [inline unittests/v1.2.0/test_drs_objects.py:17]: Should this be removed?

File: unittests/v1.2.0/test_drs_objects.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    response = MockResponse(response = expected_good_drs_object)
+
+    add_case_response_schema(test_object, schema_name, case_name, case_description, response)
+
+    assert test_object.cases[0].status == Status.PASS
+
+def test_invalid_id():
+
+    # test for missing or not string drs_id field
+    # expect status == "FAIL"
```
Comment (emre-f) [inline unittests/v1.2.0/test_drs_objects.py:40]: Should this be removed?

File: unittests/v1.2.0/test_report_runner.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            for case in test["cases"]:
+                case["start_time"] = ""
+                case["end_time"] = ""
+    return string_format(actual_final_json)
+
+def validate_timestamp_format(timestamp):
+    good_timestamp = re.search("(\d{4}-\d{2}-\d{2})T(\d{2}:\d{2}:\d{2})Z", timestamp)
+    return good_timestamp != None
+
+#def test_report_runner():
```
Comment (emre-f) [inline unittests/v1.2.0/test_report_runner.py:32]: What is this commented out function? Is it meant to be removed?

Comment (alextsaihi) [reply unittests/v1.2.0/test_report_runner.py:32]: test_report_runner() was the previous line by line, end-to-end testing so if we decide to include it instead of going by # passing tests we could

File: unittests/resources/bad_service_info_2.json
Comment (alextsaihi) [reply unittests/resources/bad_service_info_2.json:1]: this is a mystery 🧐 I didn't add this file but might've been a wonky rebase thing that happened

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
Before reviewing this PR, please rebase the "at-unittests" branch on "develop" to ensure that we only see the changes relevant to this PR. Thank you, @alextsaihi

**Comment by alextsaihi** [comment]
Hi @yash-puligundla @Chen2x , the tests that are having errors are the functions from `[report_runner.py](https://github.com/ga4gh/drs-compliance-suite/blob/at-unittests/compliance_suite/report_runner.py)` that start with
`def test_[something]`
but aside from that, it should be ready for review

**Comment by yash-puligundla** [comment]
Hi @alextsaihi,

Thank you for the update. Here are some suggestions from the first pass:

1. Remove the `nohup.out` file and add it to the `.gitignore` file to exclude it from version control.

2. In the GitHub Actions workflow, it would be beneficial to run the `good_mock_server_v1.2.0.py` script with different `auth_type` values and verify if each of the output reports meets the expected results. This can help ensure that the server behaves correctly with different authentication types.

In unittests/v1.2.0/test_report_runner.py at line 39, I noticed that you are sending requests to the good_mock_server running on port 8089 with different auth_types. However, there seems to be some confusion here. When I mentioned running the good_mock_server with different auth_types, I was referring to running the good_mock_server four different times in the GitHub Actions workflow, each time with a different auth_type. Here's an example:
```
nohup python unittests/good_mock_server_v1.2.0.py --auth_type "none" --app_host "0.0.0.0" --app_port "8089" &
nohup python unittests/good_mock_server_v1.2.0.py --auth_type "basic" --app_host "0.0.0.0" --app_port "8088" &
nohup python unittests/good_mock_server_v1.2.0.py --auth_type "bearer" --app_host "0.0.0.0" --app_port "8087" &
nohup python unittests/good_mock_server_v1.2.0.py --auth_type "passport" --app_host "0.0.0.0" --app_port "8086" &
```
By running the GM server with different auth_type values, we can test the DRS CS on each of these servers using the appropriate configuration files at `https://github.com/ga4gh/drs-compliance-suite/tree/develop/compliance_suite/config/config_samples`

Please let me know if you have any further questions or concerns.

**Comment by alextsaihi** [comment]
To whomever:

Current issue with a bundle `id: "a1dd4ae2-8d26-43b0-a199-342b64c7dff6" listed in `config_samples.json` is having `issues so unittests will produce errors. Tech team ticket [here](https://www.ebi.ac.uk/panda/jira/browse/GA4GHTT-126)

Once this issue is fixed, please come back here and uncomment the line 32 in `unittests/v1.2.0/test_report_generation.py` underneath `# USE THIS: ...` to generate new expected jsons that are used for unittesting.

File: nohup.out [hunk 4]
```diff
@@ -0,0 +1,300 @@
+WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
```
Comment (yash-puligundla) [inline nohup.out:1]: consider removing  this file using `git rm <filepath>`

File: sample.json [hunk 5]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (yash-puligundla) [inline sample.json:1]: consider removing this
`git rm <filepath>`

File: unittests/good_mock_server_v1.2.0.py [hunk 6]
```diff
@@ -19,6 +19,15 @@
 auth_basic = HTTPBasicAuth()
 auth_bearer = HTTPTokenAuth(scheme="Bearer")
 
+@auth_bearer.error_handler
```
Comment (yash-puligundla) [inline unittests/good_mock_server_v1.2.0.py:22]: TODO: yash: check where this coming from?

File: unittests/output/actual_json [hunk 7]
```diff
@@ -0,0 +1,472 @@
+{
```
Comment (yash-puligundla) [inline unittests/output/actual_json:1]: If this file is not being used anymore, consider removing this

File: unittests/output/expected_good.json [hunk 8]
```diff
@@ -0,0 +1,613 @@
+{
```
Comment (yash-puligundla) [inline unittests/output/expected_good.json:1]: If unused, remove this file

File: unittests/output/expected_good.json [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+    "platform_description": "test",
+    "input_parameters": {
+      "server_base_url": "http://localhost:8089/ga4gh/drs/v1"
+    },
+    "start_time": "",
+    "end_time": "",
+    "status": "FAIL",
+    "summary": {
+      "unknown": 0,
+      "passed": 39,
```
Comment (yash-puligundla) [inline unittests/output/expected_good.json:17]: passed: 40
failed:0

File: unittests/output/expected_good.json [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+        "phase_name": "drs object info",
+        "phase_description": "run all the tests for drs object info endpoint",
+        "start_time": "",
+        "end_time": "",
+        "status": "FAIL",
+        "summary": {
+          "unknown": 0,
+          "passed": 24,
+          "warned": 0,
+          "failed": 1,
```
Comment (yash-puligundla) [inline unittests/output/expected_good.json:93]: status: PASS
failed: 0

File: unittests/output/expected_good.json [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+                "log_messages": [],
+                "start_time": "",
+                "end_time": "",
+                "status": "PASS",
+                "message": "Content-Type matches expected type"
+              },
+              {
+                "case_name": "DRS Object Info response schema validation",
+                "case_description": "Validate DRS Object Info response schema when status = 200",
+                "log_messages": [
```
Comment (yash-puligundla) [inline unittests/output/expected_good.json:381]: log_messages: []

File: unittests/output/expected_good.json [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+                ],
+                "start_time": "",
+                "end_time": "",
+                "status": "FAIL",
+                "message": "'access_methods' is a required property"
```
Comment (yash-puligundla) [inline unittests/output/expected_good.json:387]: status: PASS
message:"Schema Validation Successful"

File: unittests/test_service_info.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
+"""
```
Comment (yash-puligundla) [inline unittests/test_service_info.py:1]: remove this file if it doesn't apply anymore

File: unittests/resources/bad_service_info_2.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (yash-puligundla) [inline unittests/resources/bad_service_info_2.json:1]: if unused, remove the file

File: unittests/utils.py [hunk 15]
```diff
@@ -1,9 +1,9 @@
 # Mimicks a HTTP request response
 class MockResponse():
-    def __init__(self, response):
+    def __init__(self, response=None, status_code=None, headers=None):
```
Comment (yash-puligundla) [inline unittests/utils.py:3]: If this is no longer true, consider reverting these changes

File: unittests/v1.2.0/test_drs_objects.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
+"""
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_drs_objects.py:1]: consider removing this file, if it is not up to date

File: unittests/v1.2.0/test_report_generation.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        # with open('unittests/output/expected_' + authtype + '.json', 'w', encoding="utf-8") as f: json.dump(good_json, f)
+
+        expect_final_json = json.loads(
+            open("unittests/output/expected_" + authtype + ".json", "r").read()
+        )
+
+        actual_json_s = str(good_json).replace("'", '"').replace("\\","")
+        expect_json_s = str(expect_final_json).replace("'", '"').replace("\\","")
+
+        # expect_final_json will vary throughout production of cs
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_report_generation.py:41]: we decided that we would add the code to check statuses from `test_validate_response` here. And also, `git rm test_validate_response.py`

File: unittests/v1.2.0/test_report_generation.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        assert actual_json_s == expect_json_s
+
+def test_send_request():
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_report_generation.py:44]: TODO: yash to add a sample test for send_request.

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
Thanks for the updates Alex. I have added some in-line comments from ourreview meeting today. Please let me know if you have any questions.

File: unittests/good_mock_server_v1.2.0.py (hunk 6 above)
Comment (yash-puligundla) [reply unittests/good_mock_server_v1.2.0.py:22]: @alextsaihi 
After investigating, I have identified the source of this change. The commit was originally added to Emre's old unittests branch, and your branch picked it up since it is based on Emre's branch. In conclusion, these code changes are appropriately located and align with the intended structure.

File: unittests/output/expected_basic.json [hunk 19]
```diff
@@ -0,0 +1,611 @@
+{
```
Comment (yash-puligundla) [inline unittests/output/expected_basic.json:1]: I recommend renaming this file to `expected_good_basic.json` and similar renaming to the rest of the files to maintain the naming convention

File: unittests/v1.2.0/test_report_generation.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
+from compliance_suite.report_runner import *
+
+auth_types = ["none", "basic", "bearer", "passport"]
+
+def test_constructor():
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_report_generation.py:5]: I suggest removing this test since it only verifies the creation of the default empty constructor, which seems unrelated to the current context of this test file. This test appears to be arbitrary and does not contribute to the purpose of the file.

File: unittests/v1.2.0/test_report_generation.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        for phase in good_json["phases"]:
+            phase["start_time"] = ""
+            phase["end_time"] = ""
+            for test in phase["tests"]:
+                test["start_time"] = ""
+                test["end_time"] = ""
+                for case in test["cases"]:
+                    case["start_time"] = ""
+                    case["end_time"] = ""
+                    assert case["status"] == "PASS"
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_report_generation.py:37]: After spending some time on this code, I recommend revising our approach. Instead of performing a blanket check for status = PASS, we should account for the specific scenario where status = SKIP. I suggest we either address the case where status = SKIP or remove this assertion altogether.

File: compliance_suite/constants.py [hunk 22]
```diff
@@ -3,7 +3,7 @@
 SERVICE_INFO_SCHEMA = "service_info.json"
 DRS_ACCESS_SCHEMA= "access_url.json"
 TESTBED_NAME = "DRS Compliance Suite"
-TESTBED_VERSION = "v0.0.0"
+TESTBED_VERSION = "v0.2.0"
```
Comment (yash-puligundla) [inline compliance_suite/constants.py:6]: The current testbed version is v1.0.3. Refer to https://github.com/ga4gh/drs-compliance-suite/tags for the latest version.
This change might also require us to update the version in the expected json reports

File: unittests/v1.2.0/test_report_generation.py [hunk 23]
```diff
@@ -0,0 +1,87 @@
+from compliance_suite.report_runner import *
```
Comment (yash-puligundla) [inline unittests/v1.2.0/test_report_generation.py:1]: Here is a suggestion to parameterize the test using different input values
```
import json
from compliance_suite.report_runner import report_runner
from parameterized import parameterized

def mock_report_runner():
    """
    This method generates and returns a list of tuples representing the good mock server report

    Each tuple contains two values:
    - authtype: Represents the authentication type.
    - good_mock_actual_report_json: Represents the report JSON generated by the mock report runner.
    """

    reports = []
    auth_types = ["none", "basic", "bearer", "passport"]
    port_numbers = ["8089", "8090", "8091", "8092"]

    for authtype, port in zip(auth_types, port_numbers):
        good_mock_actual_report = report_runner(
            server_base_url="http://localhost:" + port + "/ga4gh/drs/v1",
            platform_name="good mock server",
            platform_description=authtype,
            drs_version="1.2.0",
            config_file="compliance_suite/config/config_samples/config_" + authtype + ".json"
        )

        good_mock_actual_report_json = json.loads(good_mock_actual_report.to_json())

        # Remove timestamps
        good_mock_actual_report_json["start_time"] = ""
        good_mock_actual_report_json["end_time"] = ""
        for phase in good_mock_actual_report_json["phases"]:
            phase["start_time"] = ""
            phase["end_time"] = ""
            for test in phase["tests"]:
                test["start_time"] = ""
                test["end_time"] = ""
                for case in test["cases"]:
                    case["start_time"] = ""
                    case["end_time"] = ""

        reports.append((authtype, good_mock_actual_report_json))

    return reports

@parameterized.expand(mock_report_runner())
def test_report_runner_using_gms(authtype, good_mock_actual_report_json):
    expect_final_json = json.loads(open(f"unittests/output/expected_{authtype}.json", "r").read())
    assert good_mock_actual_report_json == expect_final_json
```

Comment (yash-puligundla) [reply unittests/v1.2.0/test_report_generation.py:1]: I unfortunately do not have the bandwidth to write unit tests for the "send_request" method. However, I highly advise you to refer to a tutorial on python unit testing if you wish to proceed with the task. [Here](https://www.digitalocean.com/community/tutorials/how-to-use-unittest-to-write-a-test-case-for-a-function-in-python) is a tutorial I recommend.

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
Hi @alextsaihi I added a few more comments and code suggestions. Please let me know if you have any further questions

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
Hi @alextsaihi I added a few more comments and code suggestions. Please let me know if you have any further questions

**Comment by yash-puligundla** [comment]
Hi @alextsaihi, thank you for the updates. After the stand-up, let's proceed with a code review. However, I did come across a small bug. The version of ga4gh-testbed-lib==1.0.3 is incorrect. I recommend using the latest version from the ga4gh-testbed-lib repository. Otherwise, the unit tests will continue to fail during the installation of this library.

**Comment by yash-puligundla** [comment]
@alextsaihi Here is the summary of our code walk through from today

- [ ] README needs to be reverted to the version from latest develop + any intentional changes made as a part of your PR (we should add instructions on how to run the unittests - i.e, the command to deploy mock servers before running the unittests and the command to terminate them once unittests run is completed)
- [ ] fix the version of ga4gh-testbed-lib in 2 files 
- [ ] If expected_reports.py is not being used, remove it from the PR
- [ ] Replace request_sample.json with the json in expected_service_info... file to avoid duplication
- [ ] Replace the code in test_report_generation.py with the suggested code in the comments to ensure we are running the tests with parameters and are able to track any assertion failures back to the tests. Follow the same structure with the send_request test

**Comment by yash-puligundla** [comment]
Thanks @alextsaihi. Here are some follow-up suggestions. Please let me know if you have any questions 
- [ ] Use `python3` instead of `python` in the command to run the unittests in the README to follow the convention
- [ ] TESTBED_VERSION should be "v1.0.3" (this refers to the version of the DRS compliance suite repository) 
ga4gh-testbed-lib should be "0.2.0" (this refers to the version of ga4gh-testbed-lib repository)  which has been correctly updated. Thank you!
- [ ] Add the temporary files created by unittests/start.sh to .gitignore
- [ ] In unittests/v1.2.0/test_report_generation.py, func: test_send_response() ensure that the JSON objects are asserted and not strings.
- [ ] In unittests/v1.2.0/test_report_generation.py, fix the indentation of the return statement in mock_request()
- [ ] In unittests/v1.2.0/test_report_generation.py, remove any unused or duplicate import statements

**Review by yash-puligundla — APPROVED** [review]
Looks Great, Thanks @alextsaihi!

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
`test_send_requests` is failing due to `Unauthorized Request`. Could you please take a look? 
https://github.com/ga4gh/drs-compliance-suite/actions/runs/5157514163/jobs/9290024914?pr=30

Everything else Looks great!

**Comment by alextsaihi** [comment]
Current bug ticket: https://www.ebi.ac.uk/panda/jira/browse/GA4GHTT-131

**Review by yash-puligundla — APPROVED** [review]
Thank you for all your work, @alextsaihi. Approving and merging the PR.
